From f127a937f7a7ef7f4917fac25a69750515f7ad52 Mon Sep 17 00:00:00 2001
From: Mohan Raj <quic_mrajraje@quicinc.com>
Date: Mon, 10 Jun 2024 18:33:13 +0530
Subject: [PATCH] hostapd: get max simultaneous link capability of mld sta

The maximum number of simultaneous link sub-field of the
STA MLD capability. Capability indicates the max number
of STAs affiliated to the STA-MLD

USage: hostapd_cli -i <interface> -l <link> all_sta

Output:

00:1a:f5:22:12:32
flags=[AUTH][ASSOC][AUTHORIZED][WMM][MFP][HT][VHT][HE][EHT]
aid=2
capability=0x0
listen_interval=1
supported_rates=8c 12 98 24 b0 48 60 6c fb
timeout_next=NULLFUNC POLL
vendor_oui=00:00:00
dot11RSNAStatsSTAAddress=00:1a:f5:22:12:32
dot11RSNAStatsVersion=1
dot11RSNAStatsSelectedPairwiseCipher=00-0f-ac-4
dot11RSNAStatsTKIPLocalMICFailures=0
dot11RSNAStatsTKIPRemoteMICFailures=0
wpa=2
AKMSuiteSelector=00-0f-ac-8
hostapdWPAPTKState=11
hostapdWPAPTKGroupState=0
hostapdMFPR=1
rx_packets=132
tx_packets=9
rx_bytes=58787
tx_bytes=2518
inactive_msec=21070
signal=66
rx_rate_info=25942
tx_rate_info=60
rx_vht_mcs_map=ffaa
tx_vht_mcs_map=ffaa
ht_mcs_bitmask=ffffffff000000000000
connected_time=224
supp_op_classes=80737475767778797a7b7c7d7e7f808182
uniibands supported:
        UNII_1 UNII_2A UNII_2C UNII_3 UNII_4
        UNII_1
        UNII_2A
        UNII_2C
        UNII_3
        UNII_4
min_txpower=0
max_txpower=75
vht_caps_info=0x3383f9b2
ht_caps_info=0x09ef
ext_capab=04000f00000000400040000000
max_nss=4
max_vhtmcs=9
max_simul_links=2

Signed-off-by: Mohan Raj <quic_mrajraje@quicinc.com>
---
 src/ap/ctrl_iface_ap.c | 14 ++++++++++++++
 1 file changed, 14 insertions(+)

diff --git a/src/ap/ctrl_iface_ap.c b/src/ap/ctrl_iface_ap.c
index 72821b4..9c7df8e 100644
--- a/src/ap/ctrl_iface_ap.c
+++ b/src/ap/ctrl_iface_ap.c
@@ -726,6 +726,20 @@ static int hostapd_ctrl_iface_sta_mib(struct hostapd_data *hapd,
 	 }
 #endif /* CONFIG_IEEE80211N */

+#ifdef CONFIG_IEEE80211BE
+	if (sta->mld_info.mld_sta == true) {
+		u16 mld_sta_capa = sta->mld_info.common_info.mld_capa;
+		u8 max_simul_links = mld_sta_capa & 0xF;
+
+		ret = os_snprintf(buf + len, buflen-len,
+				 "max_simul_links=%d\n",
+				 max_simul_links);
+
+		if (os_snprintf_error(buflen - len, ret))
+			return len;
+		len += ret;
+	}
+#endif /* CONFIG_IEEE80211BE */
 	return len;
 }

--
2.34.1

